icicle/include/icicle/program/symbol.h

```diff
@@ -0,0 +1,160 @@
+#pragma once
+
+#include <memory>
+#include <vector>
+
+namespace icicle {
+
+  /**
+   * @brief Enum to represent all the operation supported by the Symbol class.
+   *
+   */
+  enum OpCode {
+    OP_ADD = 0,
+    OP_MULT,
+    OP_SUB,
+    OP_INV,
+
+    NOF_OPERATIONS,
+
+    OP_INPUT,
+    OP_CONST,
+  };
+
+  /**
+   * @brief a node at the data-flow grapsh (DFG) that represent a single operation
+   *
+   * This class contains a data about an operation done by the Symbol class.
+   * It saves oll the data to describe the operation done and point to its operands.
+   * At any given time a Sybol contains an Operation instance that represent all the
+   * calculation history done so far to reach the current value.
+   */
+
+  template <typename S>
+  class Operation
+  {
+  public:
+    OpCode m_opcode;
+    std::shared_ptr<Operation<S>> m_operand1; // 1st operand if exist
+    std::shared_ptr<Operation<S>> m_operand2; // 2nd operand if exist
+
+    // optinal parameters:
+    std::unique_ptr<S> m_constant; // for OP_CONST: const value
+
+    // implementation:
+    int m_mem_addr; // location at the input vectors
+
+    // Constructor
+    Operation<S>(
+      OpCode opcode,
+      std::shared_ptr<Operation<S>> operand1,
+      std::shared_ptr<Operation<S>> operand2 = nullptr,
+      std::unique_ptr<S> constant = nullptr,
+      int mem_addr = -1)
+        : m_opcode(opcode), m_operand1(operand1), m_operand2(operand2), m_mem_addr(mem_addr),
+          m_constant(std::move(constant))
+    {
+    }
+
+    bool was_visited(bool set_as_visit)
```

Comment:
maybe is_visited? all this grammar in code is sometimes weird.
BTW this is reminding me something :)